File: myinterviewtest/src/main/java/com/pltest/myinterviewtest/MainActivity.java
package com.pltest.myinterviewtest;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    RecyclerView mRecycler;
    ProgressView mProgress;
    Handler handler;
    int itemWidth = 0;

    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mRecycler = findViewById(R.id.recycler);
        mProgress = findViewById(R.id.progress);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(LinearLayoutManager.HORIZONTAL);
        mRecycler.setLayoutManager(linearLayoutManager);
        List<String> list = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            list.add(" item : " + i);
        }

        MyAdapter adapter = new MyAdapter(list, new MyAdapter.OnChange() {
            @Override
            public void chagne(int value) {
                itemWidth = value;
                Log.w("tagg","itemWidth : "+itemWidth);
            }
        });

        mRecycler.setAdapter(adapter);

//        handler = new Handler() {
//            @Override
//            public void handleMessage(@NonNull Message msg) {
//                super.handleMessage(msg);
//                handler.sendEmptyMessage(1);
//                if (msg.what < 100) {
//                    mProgress.setChangeValue(msg.what);
//                    change(msg.what);
//                }
//            }
//        };

//        handler.sendEmptyMessage(1);
//        Log.w("MainActivity", "onCreate");
//        String s3 = "abc";
//        String s2 = new String("abc");
//        String s1 = "abc";
//
//        Log.w("tagg", (s1 == s2) + "");
//        Log.w("tagg", (s1 == s3) + "");
//        Log.w("tagg", (s3 == s2) + "");

        mRecycler.setOnScrollChangeListener(new View.OnScrollChangeListener() {
            @Override
            public void onScrollChange(View v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
               int x =  getScollYDistance((LinearLayoutManager) mRecycler.getLayoutManager());
                Log.w("tagg", " x  : " + x);
                float changeVa =   (x/1800f)*100f;
                mProgress.setChangeValue((int) changeVa);
            }
        });
    }

    public int getScollYDistance(LinearLayoutManager layoutManager) {
        int position = layoutManager.findFirstVisibleItemPosition();
        View firstVisiableChildView = layoutManager.findViewByPosition(position);
        int itemHeight = firstVisiableChildView.getHeight();
        return (position) * itemHeight - firstVisiableChildView.getTop();
    }

    public void change(int what) {
        handler.sendEmptyMessageDelayed(++what, 50);
    }

    public void clickAct(View v) {
        startActivity(new Intent(this, MainActivity2.class));
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        Log.w("MainActivity", "onRestart");
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.w("MainActivity", "onStart");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.w("MainActivity", "onResume");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.w("MainActivity", "onPause");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.w("MainActivity", "onStop");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.w("MainActivity", "onDestroy");
    }
}

File: demo0624/src/main/java/com/pl/demo0624/MainActivity.kt
package com.pl.demo0624

import android.os.Bundle
import android.os.Handler
import android.util.Log
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        var str = "我";

//        val str = "一"
//        val length = str.toByteArray().size
//        Log.w("tagg", length.toString() + "")
//        val list: List<Int> = ArrayList<Int>()

        var list: ArrayList<Int> = ArrayList()
        list.add(5)
        list.add(3)
        list.add(6)
        list.add(7)
        list.add(4)
        list.add(8)
        list.add(9)
        list.add(2)
        list.add(1)

//         var list = mutableListOf<Int>(5,3,6,7,4,8,9,6,2,1)
//        bubbleSort(list)
//        printList(list)
//        selectSort(list)
//        printList(list)

        printList(list)
        quickSort(list, 0, list.size-1)
        printList(list)

    }

    fun quickSort(arr: ArrayList<Int>, low: Int, high: Int) {
        if (low < high) {
            val part = partition(arr, low, high)
            quickSort(arr, low, part - 1)
            quickSort(arr, part + 1, high)
        }
    }

    private fun partition(arr: ArrayList<Int>, low: Int, high: Int): Int {
        var low = low
        var high = high
        val pivot = arr[low]
        while (low < high) {
            while (low < high && arr[high] >= pivot) {
                high--
            }
            arr[low] = arr[high]
            while (low < high && arr[low] <= pivot) {
                low++
            }
            arr[high] = arr[low]
        }
        arr[low] = pivot
        return low
    }

    fun bubbleSort(list: ArrayList<Int>) {
        for (i in list.indices) {
            for (j in 0 until list.size - i - 1) {
                if (list[j] < list[j + 1]) {
                    var temp = list[j]
                    list[j] = list[j + 1]
                    list[j + 1] = temp
                }
            }
        }

    }


    fun selectSort(list: ArrayList<Int>) {
        for (i in list.indices) {
            for (j in i until list.size) {
                if (list[i] > list[j]) {
                    var temp = list[i]
                    list[i] = list[j]
                    list[j] = temp
                }
            }
        }
    }

    fun printList(list: List<Int>) {
        for (i in list.indices) {
            Log.w("tagg", "" + i + " : " + list[i])
        }
        Log.w("tagg", "-------------------------")
    }
}
